genie-web/src/main/java/com/netflix/genie/web/services/impl/LocalJobKillServiceImpl.java

```diff
@@ -176,6 +209,28 @@ public void onKillJobEvent(@NotNull final KillJobEvent event) throws GenieExcept
         this.killJob(event.getId(), event.getReason());
     }
 
+    @Override
+    public void registerCallbackForKill(
+        @NotBlank(message = "No id entered. Unable to register job for killing.") final String id,
+        @NotNull final Runnable callback) {
+
+        jobKillCallbacks.put(id, callback);
+
+    }
+
+    @Override
+    public void deregisterCallbackForKill(
+        @NotBlank(message = "No id entered. Unable to deregister job for killing.") final String id
+    ) {
+        jobKillCallbacks.remove(id);
+    }
+
+    @VisibleForTesting
+    //Testing utility
+    Map<String, Runnable> getJobKillCallbacks() {
+        return jobKillCallbacks;
```

Comment:
At the very least I would prefer you return some sort of Unmodifiable view out of this method using `Collections` class so that if some malicious piece of code tries to use this it can't actually modify your internal state. I don't even like the idea of exposing this at all.